content/blog/2021-05-valuable.md

```diff
@@ -0,0 +1,213 @@
+event-based diagnostic information. Some consider it a structured logging
+framework, and while it can fill that use case, it can do a lot more. For
+example, [Console] aims to become a powerful tool for debugging async Rust
+applications and uses Tracing as its backbone. [Tokio] and other libraries emit
+instrumentation via Tracing. Console aggregates the events into a model of the
+application's execution, enabling the developer to gain insights into bugs and
+other issues.
+
+![Instrumentation -> trait object -> event collection](https://i.imgur.com/YJIrHK2.png)
+
+Instrumented applications emit events with rich, structured data, and collectors
+receive the events. Of course, at compile-time, the instrumented application and
+the event collectors do not know about each other. A trait object bridges the
+instrumentation half with the collection half enabling collectors to register
+themselves dynamically. So, passing rich, structured data from the
+instrumentation half to the collector requires passing it through the trait
+object boundary. Tracing supports this at a minimal level today but does not
+support passing nested data.
+
+Let's look at an actual use case. Given an HTTP service, at the start of an HTTP
+request, we want to emit a tracing event that includes relevant HTTP headers.
+The data may look something like this.
+
+```javascript=
+{
+  user_agent: "Mozilla/4.0 (compatible; MSIE5.01; Windows NT)",
+  host: "www.example.com",
+  content_type: {
+    mime: "text/xml",
+    charset: "utf-8",
+  },
+  accept_encoding: ["gzip", "deflate"],
+}
+```
+
+In the application, a Rust struct stores the headers.
+
+```rust=
+struct Headers {
+    user_agent: String,
+    host: String,
+    content_type: ContentType,
+    accept_encoding: Vec<String>,
+}
+
+struct ContentType {
+    mime: String,
+    charset: String,
+}
+```
+
+We want to pass this data to the event collector, but how? We cannot give a
+reference to the `Headers` struct. We could use a data type like
+[`serde_json::Value`][json] to pass arbitrary structured data, but this would
+require copying the data from our application's struct to hand it to the
+collector.
+
+The Valuable crate aims to solve this problem. In the HTTP header case, first,
+we would implement `Valuable` for our `Headers` type. Then, we can pass a `&dyn
+Valuable` reference to the event collector. The collector can use Valuable's API
+to inspect the value and extract data relevant to its use case. Inspecting
+Valuable instances is done using the visitor pattern. The collector implements
+`Visit` and traverses the value. If our collector wishes to extract all
+"accept-encoding" header values, the visitor will look something like this.
+
+```rust=
+// Visit the root of the Headers struct
+struct VisitHeaders(Vec<String>);
+
+// Visit the accept_encoding field.
+struct VisitAcceptEncoding<'a>(&'a mut Vec<String>);
+
+impl Visit for VisitHeaders {
```

Comment:
This example confused me a bit and took a while to understand. Both it and the text before here could probably be improved.

Reply:
I tweaked the example using comments and named fields. I tweaked the paragraph above as well.